File: identify1-recommended/ACRCloudProtocolV1.java
 import java.io.BufferedOutputStream;
 import java.io.BufferedReader;
 import java.io.ByteArrayOutputStream;
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
 import java.io.InputStreamReader;
 import java.net.HttpURLConnection;
 import java.net.URL;
 import java.util.Calendar;
 import java.util.HashMap;
 import java.util.Map;
 import javax.crypto.Mac;
 import javax.crypto.spec.SecretKeySpec;
 
 import org.apache.commons.codec.binary.Base64;
 
 public class ACRCloudProtocolV1 {
 
 	private String encodeBase64(byte[] bstr) {
 		Base64 base64 = new Base64();
 		return new String(base64.encode(bstr));
 	}
 
 	private String encryptByHMACSHA1(byte[] data, byte[] key) {
 		try {
 			SecretKeySpec signingKey = new SecretKeySpec(key, "HmacSHA1");
 			Mac mac = Mac.getInstance("HmacSHA1");
 			mac.init(signingKey);
 			byte[] rawHmac = mac.doFinal(data);
 			return encodeBase64(rawHmac);
 		} catch (Exception e) {
 			e.printStackTrace();
 		}
 		return "";
 	}
 
 	private String getUTCTimeSeconds() {  
 	    Calendar cal = Calendar.getInstance();   
 	    int zoneOffset = cal.get(Calendar.ZONE_OFFSET);   
 	    int dstOffset = cal.get(Calendar.DST_OFFSET);    
 	    cal.add(Calendar.MILLISECOND, -(zoneOffset + dstOffset));    
 	    return cal.getTimeInMillis()/1000 + "";
 	}  
 	
 	private String postHttp(String posturl, Map<String, Object> params, int timeOut) {
 		String res = "";
 		String BOUNDARYSTR = "*****2015.03.30.acrcloud.rec.copyright." + System.currentTimeMillis() + "*****";
 		String BOUNDARY = "--" + BOUNDARYSTR + "\r\n";
 		String ENDBOUNDARY = "--" + BOUNDARYSTR + "--\r\n\r\n";
 		
 		String stringKeyHeader = BOUNDARY +
                 "Content-Disposition: form-data; name=\"%s\"" +
                 "\r\n\r\n%s\r\n";
 		String filePartHeader = BOUNDARY +
                  "Content-Disposition: form-data; name=\"%s\"; filename=\"%s\"\r\n" +
                  "Content-Type: application/octet-stream\r\n\r\n";		
  
 		URL url = null;
 		HttpURLConnection conn = null;
 		BufferedOutputStream out = null;
 		BufferedReader reader = null;
 		ByteArrayOutputStream postBufferStream = new ByteArrayOutputStream();
 		try {
 			for (String key : params.keySet()) {
 				Object value = params.get(key);
 				if (value instanceof String || value instanceof Integer) {
 					postBufferStream.write(String.format(stringKeyHeader, key, (String)value).getBytes());
 				} else if (value instanceof byte[]) {
 					postBufferStream.write(String.format(filePartHeader, key, key).getBytes());
 					postBufferStream.write((byte[]) value);
 					postBufferStream.write("\r\n".getBytes());
 				}
 			}
 			postBufferStream.write(ENDBOUNDARY.getBytes());
 			
 			url = new URL(posturl);
 			conn = (HttpURLConnection) url.openConnection();
 			conn.setConnectTimeout(timeOut);
 			conn.setReadTimeout(timeOut);
 			conn.setRequestMethod("POST");
 			conn.setDoOutput(true);
 			conn.setDoInput(true);
 			conn.setRequestProperty("Accept-Charset", "utf-8");
 			conn.setRequestProperty("Content-type", "multipart/form-data;boundary=" + BOUNDARYSTR);
 
 			conn.connect();
 			out = new BufferedOutputStream(conn.getOutputStream());
 			out.write(postBufferStream.toByteArray());
 			int response = conn.getResponseCode();
 			if (response == HttpURLConnection.HTTP_OK) {
 				reader = new BufferedReader(new InputStreamReader(conn.getInputStream(), "UTF-8"));
 				String tmpRes = "";
 				while ((tmpRes = reader.readLine()) != null) {
 					if (tmpRes.length() > 0)
 						res = res + tmpRes;
 				}
 			}
 		} catch (Exception e) {
 			e.printStackTrace();
 		} finally {
 			try {
 				if (postBufferStream != null) {
 					postBufferStream.close();
 					postBufferStream = null;
 				}
 				if (out != null) {
 					out.close();
 					out = null;
 				}
 				if (reader != null) {
 					reader.close();
 					reader = null;
 				}
 				if (conn != null) {
 					conn.disconnect();
 					conn = null;
 				}
 			} catch (IOException e) {
 				e.printStackTrace();
 			}
 		}
 		return res;
 	}
 
     public String recognize(String host, String accessKey, String secretKey, byte[] queryData, String queryType, int timeout)
     {
     	String method = "POST";
     	String httpURL = "/v1/identify";
     	String dataType = queryType;
     	String sigVersion = "1";
     	String timestamp = getUTCTimeSeconds();
 
     	String reqURL = "http://" + host + httpURL;
 
     	String sigStr = method + "\n" + httpURL + "\n" + accessKey + "\n" + dataType + "\n" + sigVersion + "\n" + timestamp;
     	String signature = encryptByHMACSHA1(sigStr.getBytes(), secretKey.getBytes());
 
         Map<String, Object> postParams = new HashMap<String, Object>();
         postParams.put("access_key", accessKey);
         postParams.put("sample_bytes", queryData.length + "");
         postParams.put("sample", queryData);
         postParams.put("timestamp", timestamp);
         postParams.put("signature", signature);
         postParams.put("data_type", queryType);
         postParams.put("signature_version", sigVersion);
 
         String res = postHttp(reqURL, postParams, timeout);
 
         return res;
     }
     
 	public static void main(String[] args) {
 		File file = new File("E://sample.wav");
 		byte[] buffer = new byte[1024 * 1024];
 		if (!file.exists()) {
 			return;
 		}
 		FileInputStream fin = null;
 		int bufferLen = 0;
 		try {
 			fin = new FileInputStream(file);
 			bufferLen = fin.read(buffer, 0, buffer.length);
 		} catch (Exception e) {
 			e.printStackTrace();
 		} finally {
 			try {
 				if (fin != null) {
 					fin.close();
 				}
 			} catch (IOException e) {
 				e.printStackTrace();
 			}
 		}
 		System.out.println("bufferLen=" + bufferLen);
 		
 		if (bufferLen <= 0)
 			return;
 		
 		byte[] postDatas = new byte[bufferLen];
 		System.arraycopy(buffer, 0, postDatas, 0, bufferLen);
 		ACRCloudProtocolV1 a = new ACRCloudProtocolV1();
 		
 		String result = a.recognize("ap-southeast-1.api.acrcloud.com", "xxxxxxxx", "xxxxxxxx", postDatas, "audio", 80000);
 		System.out.println(result);
 	}
 }

File: identify protocol 1 (recommended)/ACRCloudProtocolV1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections;
using System.Net;
using System.Web;
using System.IO;
using System.Security.Cryptography;
using System.Diagnostics;

namespace ACRCloudWebAPITest
{
   class ACRCloudProtocolV1
   {
       public static string postHttp(string url, IDictionary<string, Object> postParams, int timeout)
       {
           string result = "";

           string BOUNDARYSTR = "acrcloud***copyright***2015***" + DateTime.Now.Ticks.ToString("x");
           string BOUNDARY = "--" + BOUNDARYSTR + "\r\n";
           var ENDBOUNDARY = Encoding.ASCII.GetBytes("--" + BOUNDARYSTR + "--\r\n\r\n");

           var stringKeyHeader = BOUNDARY +
                          "Content-Disposition: form-data; name=\"{0}\"" +
                          "\r\n\r\n{1}\r\n";
           var filePartHeader = BOUNDARY +
                           "Content-Disposition: form-data; name=\"{0}\"; filename=\"{1}\"\r\n" +
                           "Content-Type: application/octet-stream\r\n\r\n";

           var memStream = new MemoryStream();
           foreach (var item in postParams)
           {
               if (item.Value is string)
               {
                   string tmpStr = string.Format(stringKeyHeader, item.Key, item.Value);
                   byte[] tmpBytes = Encoding.UTF8.GetBytes(tmpStr);
                   memStream.Write(tmpBytes, 0, tmpBytes.Length);
               }
               else if (item.Value is byte[])
               {                    
                   var header = string.Format(filePartHeader, "sample", "sample");
                   var headerbytes = Encoding.UTF8.GetBytes(header);
                   memStream.Write(headerbytes, 0, headerbytes.Length);
                   byte[] sample = (byte[])item.Value;
                   memStream.Write(sample, 0, sample.Length);
                   memStream.Write(Encoding.UTF8.GetBytes("\r\n"), 0, 2);
               }
           }
           memStream.Write(ENDBOUNDARY, 0, ENDBOUNDARY.Length);

           HttpWebRequest request = null;
           HttpWebResponse response = null;
           Stream writer = null;
           StreamReader myReader = null;
           try
           {
               request = (HttpWebRequest)WebRequest.Create(url);
               request.Timeout = timeout;
               request.Method = "POST";              
               request.ContentType = "multipart/form-data; boundary=" + BOUNDARYSTR;

               memStream.Position = 0;
               byte[] tempBuffer = new byte[memStream.Length];
               memStream.Read(tempBuffer, 0, tempBuffer.Length);

               writer = request.GetRequestStream();
               writer.Write(tempBuffer, 0, tempBuffer.Length);
               writer.Close();
               writer = null;

               response = (HttpWebResponse)request.GetResponse();
               myReader = new StreamReader(response.GetResponseStream(), Encoding.UTF8);
               result = myReader.ReadToEnd();
           }
           catch (WebException e)
           {
               Debug.WriteLine("timeout:\n" + e.ToString());
           }
           catch (Exception e)
           {
               Debug.WriteLine("other excption:" + e.ToString());
           }
           finally
           {
               if (memStream != null)
               {
                   memStream.Close();
                   memStream = null;
               }
               if (writer != null)
               {
                   writer.Close();
                   writer = null;
               }
               if (myReader != null)
               {
                   myReader.Close();
                   myReader = null;
               }
               if (request != null)
               {
                   request.Abort();
                   request = null;
               }
               if (response != null)
               {
                   response.Close();
                   response = null;
               }
           }

           return result;
       }

       public static string encryptByHMACSHA1(string input, string key)
       {
           HMACSHA1 hmac = new HMACSHA1(System.Text.Encoding.UTF8.GetBytes(key));
           byte[] stringBytes = Encoding.UTF8.GetBytes(input);
           byte[] hashedValue = hmac.ComputeHash(stringBytes);
           return encodeToBase64(hashedValue);
       }

       public static string encodeToBase64(byte[] input)
       {
           string res = Convert.ToBase64String(input, 0, input.Length);
           return res;
       }

       public static string recognize(string host, string accessKey, string secretKey, byte[] queryData, string queryType, int timeout = 8000)
       {
           string method = "POST";
           string httpURL = "/v1/identify";
           string dataType = queryType;
           string sigVersion = "1";
           string timestamp = ((int)DateTime.UtcNow.Subtract(new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc)).TotalSeconds).ToString();

           string reqURL = "http://" + host + httpURL;

           string sigStr = method + "\n" + httpURL + "\n" + accessKey + "\n" + dataType + "\n" + sigVersion + "\n" + timestamp;
           string signature = encryptByHMACSHA1(sigStr, secretKey);

           var dict = new Dictionary<string, object>();
           dict.Add("access_key", accessKey);
           dict.Add("sample_bytes", queryData.Length.ToString());
           dict.Add("sample", queryData);
           dict.Add("timestamp", timestamp);
           dict.Add("signature", signature);
           dict.Add("data_type", queryType);
           dict.Add("signature_version", sigVersion);

           string res = postHttp(reqURL, dict, timeout);

           return res;
       }
   } 
   class Program
   {
       static void Main(string[] args)
       {
           using (FileStream fs = new FileStream(@"E:\sample.wav", FileMode.Open))
           {
               using (BinaryReader reader = new BinaryReader(fs))
               {
                   byte[] datas = reader.ReadBytes((int)fs.Length);
                   // Replace "xxxxxxxx" below with your project's access_key and access_secret.
                   string result = ACRCloudProtocolV1.recognize("ap-southeast-1.api.acrcloud.com", "xxxxxxxx", "xxxxxxxx", datas, "audio");
                   Console.WriteLine(result);
               }
           }
           
           Console.ReadLine();
       }
   }
}
